cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkContextProvider.java

```diff
@@ -16,57 +16,203 @@
+      final StreamCoordinatorClient streamCoordinatorClient = injector.getInstance(StreamCoordinatorClient.class);
+
+      // Use the shutdown hook to shutdown services, since this class should only be loaded from System classloader
+      // of the spark executor, hence there should be exactly one instance only.
+      // The problem with not shutting down nicely is that some logs/metrics might be lost
+      Services.chainStart(logAppenderService, zkClientService,
+                          kafkaClientService, metricsCollectionService, streamCoordinatorClient);
+      Runtime.getRuntime().addShutdownHook(new Thread() {
+        @Override
+        public void run() {
+          // The logger may already been shutdown. Use System.out/err instead
+          System.out.println("Shutting SparkClassLoader services");
+          Future<List<ListenableFuture<Service.State>>> future = Services.chainStop(logAppenderService,
+                                                                                    streamCoordinatorClient,
+                                                                                    metricsCollectionService,
+                                                                                    kafkaClientService,
+                                                                                    zkClientService);
+          try {
+            List<ListenableFuture<Service.State>> futures = future.get(5, TimeUnit.SECONDS);
+            System.out.println("SparkClassLoader services shutdown completed: " + futures);
+          } catch (Exception e) {
+            System.err.println("Exception when shutting down services");
+            e.printStackTrace(System.err);
+          }
+        }
+      });
+
+      // Create the context object
+      sparkContext = new ExecutionSparkContext(
+        contextConfig.getSpecification(), contextConfig.getProgramId(), contextConfig.getRunId(),
+        classLoader, contextConfig.getLogicalStartTime(), contextConfig.getArguments(),
+        contextConfig.getTransaction(), injector.getInstance(DatasetFramework.class),
+        injector.getInstance(DiscoveryServiceClient.class), metricsCollectionService, hConf,
+        injector.getInstance(StreamAdmin.class)
+      );
+      return sparkContext;
+    } catch (Exception e) {
+      throw Throwables.propagate(e);
     }
-    return context;
   }
 
-  private synchronized AbstractSparkContextBuilder getBuilder(CConfiguration conf) {
-    if (contextBuilder == null) {
-      String mrFramework = hConf.get(MRConfig.FRAMEWORK_NAME, MRConfig.LOCAL_FRAMEWORK_NAME);
-      if ("local".equals(mrFramework)) {
-        contextBuilder = new InMemorySparkContextBuilder(conf);
-      } else {
-        throw new RuntimeException("Spark does not run in distributed mode right now");
+  private static CConfiguration createCConf() throws MalformedURLException {
+    CConfiguration cConf = CConfiguration.create();
```

Comment:
is this common across other program contexts? Perhaps we should add a CConfiguration.create(URL ... resources)?